config: learn to read subzones

benesch opened the pull request:
Ignore the first four commits; those are in #19315.

----

As described in the partitioning RFC, teach the config package how to
interpret subzones stored in a ZoneConfig. In the interest of minimizing
review burden, this commit does not introduce any codepaths outside of
tests that create ZoneConfigs with subzones.

benesch committed:
config: learn to read subzones

As described in the partitioning RFC, teach the config package how to
interpret subzones stored in a ZoneConfig. In the interest of minimizing
review burden, this commit does not introduce any codepaths outside of
tests that create ZoneConfigs with subzones.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19360)
<!-- Reviewable:end -->

benesch commented:
Ok, I've rebased out #19315. RFAL!

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the []byte{} vs nil issue. i'm confused by that, so if we stick with what you have, it deserves code comments

---

Review status: 0 of 11 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/config/config.go, line 320 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8shnwSPP-GWVibkQV:-Kx8shnwSPP-GWVibkQW:b-1icft8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.go#L320)):*
> ```Go
> 		// An unset EndKey implies Key.PrefixEnd().
> 		endKeyMatches := (s.EndKey == nil && bytes.HasPrefix(keySuffix, s.Key)) || s.EndKey.Compare(keySuffix) > 0
> 		if startKeyMatches && endKeyMatches {
> ```

no lazy eval here is unfortunate. i know the formatting sucks, but wanna inline these in the if?

---

*[pkg/config/config.proto, line 83 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8tFcBGSBYG5FV0JTT:-Kx8tFcBGSBYG5FV0JTU:b-jhzy9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L83)):*
> ```protobuf
>   optional Constraints constraints = 6 [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"constraints,flow\""];
> 
>   // subzones stores config overrides for "subzones", each of which represents
> ```

looks like the nearby style is to write this comment starting with the generated go identifier, so Subzones. here and elsewhere

---

*[pkg/config/config.proto, line 85 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8u8HqGIGgrB16fL95:-Kx8u8HqGIGgrB16fL96:b-afalew) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L85)):*
> ```protobuf
>   // subzones stores config overrides for "subzones", each of which represents
>   // either an index or a partition of an index. Subzones are not applicable
>   // when the zone does not represent a table.
> ```

nit: `does not represent a table (e.g. a database or the cluster default).`

also maybe include something like Subzones are mapped to keyspans by the SubzoneSpans field, or something like that

---

*[pkg/config/config.proto, line 88 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8uYSLnpeqz0PWjAqp:-Kx8uYSLnpeqz0PWjAqq:br0eynf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L88)):*
> ```protobuf
>   repeated Subzone subzones = 8 [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"-\""];
> 
>   // subzone_spans maps each key span in a subzone to the index of an entry in
> ```

nit: to the array index

(or slice index)

---

*[pkg/config/config.proto, line 97 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8uxATuBOKeo7_CprB:-Kx8uxATuBOKeo7_CprC:b-kxm44k) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L97)):*
> ```protobuf
>   option (gogoproto.equal) = true;
> 
>   // index_id is the ID of the index that the subzone represents.
> ```

of the sql table index, or something like that

---

*[pkg/config/config.proto, line 102 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8v41VB41xwk__ZxG1:-Kx8v41VB41xwk__ZxG2:b-yo15ez) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L102)):*
> ```protobuf
>   // It is empty when the subzone represents the entire index.
>   optional string partition_name = 2 [(gogoproto.nullable) = false];
>   optional ZoneConfig config = 3 [(gogoproto.nullable) = false];
> ```

mention explictly that this ZoneConfig will never include Subzones or SubzoneSpans

---

*[pkg/config/config.proto, line 108 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8vCQT2Ihmpvip_JTF:-Kx8vCQT2Ihmpvip_JTG:b-7po417) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L108)):*
> ```protobuf
>   option (gogoproto.equal) = true;
> 
>   optional roachpb.Span span = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
> ```

i'm divided on whether to actually use roachpb.Span for this. it's not a real span, but I guess the pretty printing logic does happen to work, more or less. I think I'd be more comfortable if this were two fields, for Key and EndKey. we can make them both roachpb.Keys to start but stick a comment that if someone needs to make them plain []byte in the future, that's okay

---

*[pkg/config/config.proto, line 109 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8v8y9CpkEUSxFkLaH:-Kx8v8y9CpkEUSxFkLaI:bf2mjko) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L109)):*
> ```protobuf
> 
>   optional roachpb.Span span = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
>   optional int32 subzone_index = 2 [(gogoproto.nullable) = false];
> ```

comment plz

---

*[pkg/sql/config.go, line 84 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx9-Fe2Yqa9zD8-2UAw:-Kx9-Fe2Yqa9zD8-2UAx:b-d16nd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config.go#L84)):*
> ```Go
> 	// we were trying to retrieve (avoid infinite recursion).
> 	if id != keys.RootNamespaceID {
> 		return getZoneConfig(keys.RootNamespaceID, []byte{}, get)
> ```

why all the `[]byte{}` instead of `nil`? are you sure there's no allocation difference? (i don't recall one way or the other offhand)

---

*[pkg/sql/config_test.go, line 216 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8zbsieuR7JgSU5boV:-Kx8zbsieuR7JgSU5boW:bb7erio) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L216)):*
> ```Go
> 	//   tb1: true
> 	//   tb2: false
> 	// db1: false
> ```

db2?

---

*[pkg/sql/config_test.go, line 218 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8wFmXgFXvPl6dFiSQ:-Kx8wFmXgFXvPl6dFiSR:b-ntqnwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L218)):*
> ```Go
> 	// db1: false
> 	//   tb1: true
> 	//     p1: false [1, 1), [3, 4)
> ```

all your end keys here (except 255) are off by one

---

*[pkg/sql/config_test.go, line 285 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx9-aGhm4Ysn1ILwtUt:-Kx9-aGiCQ_oLA_5IckB:bvs0vb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L285)):*
> ```Go
> 		{tb12, []byte{42}, db1Cfg},
> 		{tb21, nil, tb21Cfg},
> 		{tb21, []byte{}, tb21Cfg},
> ```

isn't this the same as the last case?

---

*[pkg/sql/config_test.go, line 290 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx9-ryzXQaY_GDoz_-a:-Kx9-ryzXQaY_GDoz_-b:b-4c6jje) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L290)):*
> ```Go
> 		{tb21, []byte{1, 255}, p211Cfg},
> 		{tb21, []byte{2}, tb21Cfg},
> 		{tb21, []byte{3}, p212Cfg},
> ```

add a case for []byte{4}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-:-Kx905PHq9xx2nSRHhOm:b-kii0vc)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/config/config.go, line 320 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8shnwSPP-GWVibkQV:-KxBXE7UZp4HGlFB23fe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.go#L320)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

no lazy eval here is unfortunate. i know the formatting sucks, but wanna inline these in the if?
</blockquote></details>

Done.

---

*[pkg/config/config.proto, line 83 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8tFcBGSBYG5FV0JTT:-KxBZWhmhNNRdLB75ZcR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L83)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

looks like the nearby style is to write this comment starting with the generated go identifier, so Subzones. here and elsewhere
</blockquote></details>

Done.

---

*[pkg/config/config.proto, line 85 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8u8HqGIGgrB16fL95:-KxB_-3XeypBkMZFnwe9:b-pvxatm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L85)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: `does not represent a table (e.g. a database or the cluster default).`

also maybe include something like Subzones are mapped to keyspans by the SubzoneSpans field, or something like that
</blockquote></details>

Done. I left out the bit about the mapping because technically `Subzones` describes what keys it applies to, provided you have a TableDescriptor handy. I added an note to`SubzoneSpans` explaining *why* we denormalize, though.

---

*[pkg/config/config.proto, line 88 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8uYSLnpeqz0PWjAqp:-KxBatphkp2ai94Zryz_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L88)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: to the array index

(or slice index)
</blockquote></details>

Done.

---

*[pkg/config/config.proto, line 97 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8uxATuBOKeo7_CprB:-KxBb41Jy1wUdlLY5pTB:b-2b6stt) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L97)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

of the sql table index, or something like that
</blockquote></details>

Done, here and above.

---

*[pkg/config/config.proto, line 102 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8v41VB41xwk__ZxG1:-KxBbDCme95H9q4Ry38X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L102)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

mention explictly that this ZoneConfig will never include Subzones or SubzoneSpans
</blockquote></details>

Done.

---

*[pkg/config/config.proto, line 108 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8vCQT2Ihmpvip_JTF:-KxBcpGxc2rGILeHwQXd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L108)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

i'm divided on whether to actually use roachpb.Span for this. it's not a real span, but I guess the pretty printing logic does happen to work, more or less. I think I'd be more comfortable if this were two fields, for Key and EndKey. we can make them both roachpb.Keys to start but stick a comment that if someone needs to make them plain []byte in the future, that's okay
</blockquote></details>

Done.

---

*[pkg/config/config.proto, line 109 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8v8y9CpkEUSxFkLaH:-KxBcogp5KUz_7KE5W9i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/config/config.proto#L109)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

comment plz
</blockquote></details>

Done.

---

*[pkg/sql/config.go, line 84 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx9-Fe2Yqa9zD8-2UAw:-KxBZ4o5tV9GSZ6WRWGn:byeob2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config.go#L84)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

why all the `[]byte{}` instead of `nil`? are you sure there's no allocation difference? (i don't recall one way or the other offhand)
</blockquote></details>

Heh, I thought it was clearer than passing `nil`, but clearly not. Switched to `nil`.

---

*[pkg/sql/config_test.go, line 216 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8zbsieuR7JgSU5boV:-KxBXmZR5Ydhy27DMG7f:b-po2n98) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L216)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

db2?
</blockquote></details>

Whoops, thanks. Done.

---

*[pkg/sql/config_test.go, line 218 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx8wFmXgFXvPl6dFiSQ:-KxBXw8sO4gla03i0fUa:bo7r87p) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L218)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

all your end keys here (except 255) are off by one
</blockquote></details>

D'oh! Thanks. Fixed.

---

*[pkg/sql/config_test.go, line 285 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx9-aGhm4Ysn1ILwtUt:-KxBYhFFEi58rUEhzv9x:b-fieq2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L285)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

isn't this the same as the last case?
</blockquote></details>

I like testing `nil` and `[]byte{}` separately—it catches cases where the logic specifies `!= nil` when it really should specify `len != 0`.

---

*[pkg/sql/config_test.go, line 290 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx9-ryzXQaY_GDoz_-a:-KxBYgWGhMsFAMPsdyG2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L290)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

add a case for []byte{4}
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 4 of 4 files at r1, 8 of 11 files at r5, 2 of 13 files at r6, 9 of 9 files at r7.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/config_test.go, line 285 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360#-Kx9-aGhm4Ysn1ILwtUt:-KxDgj7Tz7cgdFswYUmG:b-pfq32o) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdf5cc689a8dc034bab588d10c5db63078debdae/pkg/sql/config_test.go#L285)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I like testing `nil` and `[]byte{}` separately—it catches cases where the logic specifies `!= nil` when it really should specify `len != 0`.
</blockquote></details>

ah, gotcha

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19360)*
<!-- Sent from Reviewable.io -->

benesch commented:
TFTRs!
